ta/pkcs11/src/processing_asymm.c

```diff
@@ -125,6 +129,9 @@ pkcs2tee_algorithm(uint32_t *tee_id, uint32_t *tee_hash_id,
 	case PKCS11_CKM_ECDSA_SHA512:
 		rc = pkcs2tee_algo_ecdsa(tee_id, proc_params, obj);
 		break;
+	case PKCS11_CKM_ECDH1_DERIVE:
```

Comment:
Missing  `case PKCS11_CKM_ECDH1_COFACTOR_DERIVE:` here to support the mechanism.
Note GP TEE Internal API only consider the cofactor deriveration scheme, so both `PKCS11_CKM_ECDH1_DERIVE` and PKCS11_CKM_ECDH1_COFACTOR_DERIVE` end in a cofactor derivation scheme (IIUC).

Reply:
I suppose TEE internal API method is defined with:
```
TEE_ALG_ECDH_DERIVE_SHARED_SECRET - NIST SP800-56A, Cofactor Static Unified Model
```

When I have a look at said specification there is following defined:
- 5.7.1.1 Finite Field Cryptography Diffie-Hellman (FFC DH) Primitive
- 5.7.1.2 Elliptic Curve Cryptography Cofactor Diffie-Hellman (ECC CDH) 
Primitive
- 5.7.2.1 Finite Field Cryptography MQV (FFC MQV) Primitive
- 5.7.2.2 ECC MQV Associate Value Function
- 5.7.2.3 Elliptic Curve Cryptography MQV (ECC MQV) Primitive

Then we have in PKCS#11:

- 2.3.9 Elliptic curve Diffie-Hellman key derivation (CKM_ECDH1_DERIVE)
- 2.3.10 Elliptic curve Diffie-Hellman with cofactor key derivation (CKM_ECDH1_COFACTOR_DERIVE)
- 2.3.11 Elliptic curve Menezes-Qu-Vanstone key derivation (CKM_ECMQV_DERIVE)

These all refer to ANSI X9.63.

I am not in anyway a math expert but just looking at those documents I would say that:
- TEE_ALG_ECDH_DERIVE_SHARED_SECRET would match to NIST SP800-56A/5.7.1.2 
- and that would match to CKM_ECDH1_COFACTOR_DERIVE.

Also looking at different variants described in NIST SP800-56A I would say that 5.7.1.1 and 5.7.1.2 are different methods.

That being said I believe CKM_ECDH1_DERIVE would match to NIST SP800-56A/5.7.1.1.

Now this code in here assumes that both CKM_ECDH1_DERIVE and CKM_ECDH1_COFACTOR_DERIVE are calculated the same.

Now I may be missing something but to me that is wrong.

Reply:
If I understood correctly the cofactor comes from domain parameters -- eg. from curve:
```
$ openssl ecparam -noout -text -param_enc explicit -name P-192
...
Cofactor:  1 (0x1)
...
$ openssl ecparam -noout -text -param_enc explicit -name P-224
...
Cofactor:  1 (0x1)
...
$ openssl ecparam -noout -text -param_enc explicit -name P-256
...
Cofactor:  1 (0x1)
...
$ openssl ecparam -noout -text -param_enc explicit -name P-384
...
Cofactor:  1 (0x1)
...
$ openssl ecparam -noout -text -param_enc explicit -name P-521
...
Cofactor:  1 (0x1)
```

So as we currently support only those curves then the `CKM_ECDH1_DERIVE` and `CKM_ECDH1_COFACTOR_DERIVE` would give same result. But when we would support some other curves then this would become a bigger issue?

Reply:
OP-TEE's both libtomcrypt and mbedtls variants seems to be implementing `CKM_ECDH1_DERIVE` without cofactor and not the one with cofactor as one would believe based on Global Platform spec.

Reply:
OP-TEE currently implements only the NIST curves (`TEE_ECC_CURVE_NIST_P{192,224,256,384,521}`) with cofactor 1. SM2 is also supported but only for its specific key exchange protocol (SM2 KEP) and signature algorithm (SM2 DSA).

You can find ECDH tests in `xtest 4009`: see [here](https://github.com/OP-TEE/optee_test/blob/3.16.0/host/xtest/regression_4000.c#L4793) and [here](https://github.com/OP-TEE/optee_test/blob/3.16.0/host/xtest/regression_4000_data.h#L6481).